pkg/apis/pxc/v1/pxc_types.go

```diff
@@ -508,53 +496,50 @@ func (cr *PerconaXtraDBCluster) CheckNSetDefaults(serverVersion *ServerVersion)
 		if len(serverVersion.Platform) > 0 {
 			cr.Spec.Platform = serverVersion.Platform
 		} else {
-			cr.Spec.Platform = PlatformKubernetes
+			cr.Spec.Platform = version.PlatformKubernetes
 		}
 	}
 
 	cr.setSecurityContext()
 
-	return changed, nil
+	return CRVerChanged || changed, nil
 }
 
 // setVersion sets the API version of a PXC resource.
-// The new (semver-matching) version is determined either by the CR's API version or an API version specified via the CR's annotations.
-// If the CR's API version is an empty string, it returns "v1"
-func (cr *PerconaXtraDBCluster) setVersion() error {
-	apiVersion := cr.APIVersion
+// The new (semver-matching) version is determined either by the CR's API version or an API version specified via the CR's fields.
+// If the CR's API version is an empty string and last-applied-configuration from k8s is empty, it returns current operator version.
+func (cr *PerconaXtraDBCluster) setVersion() (bool, error) {
+	if len(cr.Spec.CRVersion) > 0 {
+		return false, nil
+	}
+	apiVersion := version.Version
 	if lastCR, ok := cr.Annotations["kubectl.kubernetes.io/last-applied-configuration"]; ok {
 		var newCR PerconaXtraDBCluster
 		err := json.Unmarshal([]byte(lastCR), &newCR)
 		if err != nil {
-			return errors.Wrap(err, "unmarshal cr")
+			return false, errors.Wrap(err, "unmarshal cr")
+		}
+		if len(newCR.APIVersion) > 0 {
+			apiVersion = strings.Replace(strings.TrimPrefix(newCR.APIVersion, "pxc.percona.com/v"), "-", ".", -1)
 		}
-		apiVersion = newCR.APIVersion
-	}
-	crVersion := strings.Replace(strings.TrimPrefix(apiVersion, "pxc.percona.com/v"), "-", ".", -1)
-	if len(crVersion) == 0 {
-		crVersion = "v1"
-	}
-	version, err := v.NewVersion(crVersion)
-	if err != nil {
-		return errors.Wrap(err, "new version")
 	}
-	cr.version = version
 
-	return nil
+	cr.Spec.CRVersion = apiVersion
+	return true, nil
 }
 
 func (cr *PerconaXtraDBCluster) Version() *v.Version {
-	return cr.version
+	return v.Must(v.NewVersion(cr.Spec.CRVersion))
```

Comment:
Can we do that in `setVersion` and here just return created version object

Reply:
The problem is that we can't store external structs inside our cr, since it breaks code generation for our types.
That's why we store it as a string(and now it's a part of cr)